File: haskell/p17.hs
-- P17 Split a list into two parts; the length of the first part is given.

-- Using predefined function
f1 :: Int -> [a] -> ([a], [a])
f1 = splitAt

-- Using "take" and "drop"
f2 :: Int -> [a] -> ([a], [a])
f2 n xs = (take n xs, drop n xs)

-- Recursion with a counter
f3 :: Int -> [a] -> ([a], [a])
f3 _ [] = ([], [])
f3 0 xs = ([], xs)
f3 n (x:xs) = let (a, b) = f3 (pred n) xs in (x : a, b)

-- Tail recursion with a counter
f4 :: Int -> [a] -> ([a], [a])
f4 = f []
    where f acc _ [] = (reverse acc, [])
          f acc 0 xs = (reverse acc, xs)
          f acc n (x:xs) = f (x : acc) (pred n) xs

File: java/P17.java
// P17 Split a list into two parts; the length of the first part is given.

import core.List;
import core.Pair;

import static core.List.*;
import static core.Pair.pair;

class P17 {
    // Using 'predefined' function
    <T> Pair<List<T>, List<T>> f1(int n, List<T> list) {
        return list.splitAt(n);
    }

    // Using "take" and "drop"
    <T> Pair<List<T>, List<T>> f2(int n, List<T> list) {
        return pair(list.take(n), list.drop(n));
    }

    // Recursion with a counter
    <T> Pair<List<T>, List<T>> f3(int n, List<T> list) {
        if (list.isEmpty())
            return pair(List.<T>nil(), List.<T>nil());

        if (n == 0)
            return pair(List.<T>nil(), list);

        Pair<List<T>, List<T>> p = f3(n - 1, list.tail());

        return pair(cons(list.head(), p.getA()), p.getB());
    }

    // Tail recursion with a counter
    <T> Pair<List<T>, List<T>> f4(int n, List<T> list) {
        return f4_inner(List.<T>nil(), n, list);
    }

    private <T> Pair<List<T>, List<T>> f4_inner(List<T> acc, int n, List<T> list) {
        return list.isEmpty()
                ? pair(acc.reverse(), List.<T>nil())
                : n == 0 ? pair(acc.reverse(), list) : f4_inner(cons(list.head(), acc), n - 1, list.tail());
    }
}